Create cache dir

JohanByttner opened the pull request:
### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew/pulls) for the same update/change?

_You can erase any parts of this template not applicable to your Pull Request._

### Changes to Homebrew's Core:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031) if you'd like one.
- [x] Have you successfully ran `brew tests` with your changes locally?

### Submission

#### Change
Making Homebrew create the cache directory (FormulaLock::LOCKDIR) if it does not exist, before locking the .brewing file. This prevents a "permission denied" crash.

#### Reason
When the Caches directory does not exist, brew crashes unexpectedly. At our school, the computers are automatically configured to put the brew cache at ~/Library/Caches/Homebrew. We often change computers and thus lose our ~/Library/Caches directory. This fix is small and makes it possible to move system without having to manually re-configure brew (or more likely giving up on using it, as it just crashes suddenly).

#### Tests
Added test to check if brew locks .brewing files correctly if the cache is a nonexistent directory.
Note: Because LOCKDIR is defined as a constant, the test requires calling the private method :remove_const and then restoring LOCKDIR to its original value after the test. If there is a better way to test this, I am happy to change the code.

#### Status
Tests: Passing on local computer

bfontaine commented on Library/Homebrew/formula_lock.rb:
```diff
@@ -35,6 +36,7 @@ def with_lock
 
   def get_or_create_lockfile
     if @lockfile.nil? || @lockfile.closed?
+      FileUtils.mkdir_p(LOCKDIR) unless File.directory?(LOCKDIR)
```
`mkdir_p` already checks if the directory exists so we don’t need to test it ourselves.

bfontaine commented on Library/Homebrew/test/test_formula_lock.rb:
```diff
@@ -19,4 +19,16 @@ def test_locking_file_with_existing_lock_raises_error
+    FormulaLock::LOCKDIR.children.each(&:unlink)
+    FormulaLock.send(:remove_const, :LOCKDIR)
+    FormulaLock.const_set(:LOCKDIR, @tmp_lockdir)
+  end
```
You might want to use `ensure` to, well, _ensure_ stuff is correctly reset if the test fails:

```rb
def test_locking_file_in_nonexistent_cache_directory
  # the test goes here
ensure
  # the code that resets things goes here
end
```

See [`test_integration_cmds.rb`](https://github.com/Homebrew/homebrew/blob/8a26671c16096d8b6b8bc51f76a3b21643cbd4af/Library/Homebrew/test/test_integration_cmds.rb#L327-L339) for some examples.

The issue right now is that if e.g. the first `assert_nothing_raised` fails, the test stops here and the rest of the code is not executed.

JohanByttner commented:
I added the ensure statement after the asserts. Asserts are mostly there to warn you if the mkdir_p gets removed in the future.

I also removed the extra check on mkdir_p.

apjanke commented:
This might be tricky. The `HOMEBREW_CACHE_FORMULA` is produced from `HOMEBREW_CACHE_FORMULA = HOMEBREW_CACHE+"Formula"`. If I understand the code correctly, that's not a normal Pathname object, but a [special object with custom directory creation behavior](https://github.com/Homebrew/homebrew/blob/27524770e4a13991ceb22a17c1f4660d8a7025dc/Library/Homebrew/config.rb#L11-L18). (I'm not really sure if that singleton method gets propagated to derived objects when you do `/` on them.)

So calling `LOCKDIR = HOMEBREW_CACHE_FORMULA ... FileUtils.mkdir_p(LOCKDIR)` might bypass the cache dir creation logic that the normal `HOMEBREW_CACHE_FORMULA.mkpath` call runs. `LOCKDIR.mkpath` might be acceptable?

Maybe this code should go in the same `Pathname` extension that `cache` uses.

JohanByttner commented:
I had a look at it and it seems like the issue is with [the first few lines](https://github.com/Homebrew/homebrew/blob/27524770e4a13991ceb22a17c1f4660d8a7025dc/Library/Homebrew/config.rb#L2-L3), when the env has defined HOMEBREW_CACHE as a folder that does not exist. 

mkpath is an alias to mkdir_p and should not cause any issues. Maybe we should define mkpath as a function on the Pathname returned from [line three in config.rb](https://github.com/Homebrew/homebrew/blob/27524770e4a13991ceb22a17c1f4660d8a7025dc/Library/Homebrew/config.rb#L3). mkpath eventually gets called as ```LOCKDIR.mkpath```  (implicitly ```HOMEBREW_CACHE_FORMULA.mkpath```) in  [formula_lock](https://github.com/Homebrew/homebrew/blob/8c6777b2b8e847d5f2b447eae532fa8ecffa6467/Library/Homebrew/formula_lock.rb#L13). But if the mkpath method is not defined the lock proceeds anyway and crashes brew.

Currently mkpath does not change any permissions or ownership if the directories do exist, so there are several possible strategies. But moving the mkpath function to config.rb would make it clearer for future maintainers what is going on (looking for input here).

This was what got me the original solution to the problem:
http://stackoverflow.com/questions/17397544/brew-install-permission-denied-after-chown-r-whoami
Should I open a bug report?

JohanByttner commented:
@apjanke I moved the mkpath to config.rb, and reused the already existing declaration there. If the user ```~/Library/Caches/Homebrew``` exists, mkpath will still be called, however mkpath will not do anything as the path is already there. It should also make it clear for future maintainers what is going on.

I tested it with a few installs and it works nicely. Unit tests are also fine.

mikemcquaid commented:
@xu-cheng @UniqMartin @ilovezfs Any thoughts on this?

UniqMartin commented:
I'm not sure I understand how the current patch changes anything with respect to the original issue (the cache directory not existing). If I'm reading the code correctly, the `cache` method is still always returning a `Pathname` object and doesn't automatically call `mkpath` on it.

However, it changes things in the sense that **if** `mkpath` gets called on the cache path, it will be always created with `0775` permissions, no matter where it is located. I think it makes sense to do this for `/Library/Caches/Homebrew` (though there have been debates about the security implications of this), but it's certainly the wrong thing to do for arbitrary cache directories. I sure wouldn't want a group-writeable and world-readable cache directory to be created, if it was located in my home directory.

Thus :-1: until someone can enlighten me on the former and convince me on the latter.

JohanByttner commented:
I'm closing this, as for some reason Brew has stopped giving me an error.  Thus the extra code seems redundant (maybe there was something wrong with my initial config).

@UniqMartin On a side note, Brew does create world-reachable directories if you set ```HOMEBREW_CACHE```. Permissions then become 0755 for ```HOMEBREW_CACHE``` and ```HOMEBREW_CACHE_FORMULA``` (```{formula}.brewing``` are given permissions 644). If this is a concern, maybe we should look over how [LOCKDIR.mkpath](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/formula_lock.rb#L13) is called.

ilovezfs commented:
@JohanByttner Yes, I was having trouble reproducing it at all to test this. If you do figure out what fixed it, it would be nice to know.
